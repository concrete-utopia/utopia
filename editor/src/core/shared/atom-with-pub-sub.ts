import * as PubSub from 'pubsub-js'
import React from 'react'
import { unstable_batchedUpdates } from 'react-dom'
import { HMR, IS_JEST_ENVIRONMENT } from '../../common/env-vars'
import { useForceUpdate } from '../../components/editor/hook-utils'

// From https://github.com/dai-shi/use-context-selector/blob/2dd334d727fc3b4cbadf7876b6ce64e0c633fd25/src/index.ts#L25
const isSSR =
  typeof window === 'undefined' ||
  // eslint-disable-next-line @typescript-eslint/strict-boolean-expressions
  /ServerSideRendering/.test(window.navigator && window.navigator.userAgent) ||
  IS_JEST_ENVIRONMENT

export const useIsomorphicLayoutEffect = isSSR ? React.useEffect : React.useLayoutEffect

export interface AtomWithPubSub<T> {
  key: string
  currentValue: T
  Provider: React.FunctionComponent<React.PropsWithChildren<React.PropsWithChildren<{ value: T }>>>
}

const GlobalAtomMap: { [key: string]: AtomWithPubSub<any> } = {}

export function atomWithPubSub<T>(options: { key: string; defaultValue: T }): AtomWithPubSub<T> {
  const { key, defaultValue } = options
  if (key in GlobalAtomMap && !HMR) {
    throw new Error(`Tried to create multiple atoms with the same key: ${key}`)
  }
  const newAtom: AtomWithPubSub<T> = {
    key: key,
    currentValue: defaultValue,
    Provider: ({ children, value }) => {
      const updateAtomSynchronously = usePubSubAtomWriteOnlyInner(newAtom, 'sync')
      newAtom.currentValue = value // TODO this is sneaky and we should use API instead
      useIsomorphicLayoutEffect(() => {
        unstable_batchedUpdates(() => {
          updateAtomSynchronously(() => value)
        })
      })
      return React.createElement(React.Fragment, {}, children)
    },
  }
  GlobalAtomMap[key] = newAtom
  return newAtom
}

export function useSubscribeToPubSubAtom<T>(
  atom: AtomWithPubSub<T>,
  referentiallyStableCallback: (newData: T) => void,
): void {
  const pubsubCallback = React.useCallback(
    (
      message: string,
      data: any, // TODO once eslint for hooks is updated, replace data: any with data: T
    ) => {
      referentiallyStableCallback(data)
    },
    [referentiallyStableCallback],
  )

  React.useEffect(() => {
    const token = PubSub.subscribe(atom.key, pubsubCallback)
    return function cleanup() {
      PubSub.unsubscribe(token)
    }
  }, [atom.key, pubsubCallback])
}

export const AlwaysTrue = (): boolean => true
export const AlwaysFalse = (): boolean => false

export function usePubSubAtomReadOnly<T>(
  atom: AtomWithPubSub<T>,
  shouldUpdateCallback: (newValue: T) => boolean,
): T {
  const forceUpdate = useForceUpdate()
  const previousValueRef = React.useRef(atom.currentValue)

  const shouldUpdateCallbackRef = React.useRef(shouldUpdateCallback)
  shouldUpdateCallbackRef.current = shouldUpdateCallback

  useSubscribeToPubSubAtom(
    atom,
    React.useCallback(() => {
      if (
        previousValueRef.current !== atom.currentValue &&
        shouldUpdateCallbackRef.current(atom.currentValue)
      ) {
        forceUpdate()
      }
    }, [forceUpdate, atom]),
  )
  previousValueRef.current = atom.currentValue
  return atom.currentValue
}

function usePubSubAtomWriteOnlyInner<T>(
  atom: AtomWithPubSub<T>,
  publishMode: 'sync' | 'async',
): (newValueOrUpdater: T | ((oldValue: T) => T)) => void {
  return React.useCallback(
    (newValueOrUpdater: T | ((oldValue: T) => T)) => {
      let newValue: T
      if (typeof newValueOrUpdater === 'function') {
        // if the new value is a function, we assume it is an updater
        newValue = (newValueOrUpdater as (oldValue: T) => T)(atom.currentValue)
      } else {
        newValue = newValueOrUpdater
      }
      atom.currentValue = newValue
      if (publishMode === 'sync') {
        PubSub.publishSync(atom.key, newValue)
      } else {
        PubSub.publish(atom.key, newValue)
      }
    },
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [atom.key],
  )
}

export function usePubSubAtomWriteOnly<T>(
  atom: AtomWithPubSub<T>,
): (newValueOrUpdater: T | ((oldValue: T) => T)) => void {
  return usePubSubAtomWriteOnlyInner(atom, 'async')
}

export function usePubSubAtom<T>(
  atom: AtomWithPubSub<T>,
): [T, (newValueOrUpdater: T | ((oldValue: T) => T)) => void] {
  return [usePubSubAtomReadOnly(atom, AlwaysTrue), usePubSubAtomWriteOnly(atom)]
}
