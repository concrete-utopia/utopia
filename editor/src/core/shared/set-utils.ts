// Should allow us to guard against a type being refactored into something
// which Set allows but gives us nonsense results.
export function emptySet<T extends string | boolean | number | null | undefined>(): Set<T> {
  return new Set()
}

export function setsEqual<T>(first: Set<T>, second: Set<T>): boolean {
  if (first === second) {
    return true
  } else {
    if (first.size === second.size) {
      for (const firstValue of first) {
        if (!second.has(firstValue)) {
          return false
        }
      }
      return true
    } else {
      return false
    }
  }
}
