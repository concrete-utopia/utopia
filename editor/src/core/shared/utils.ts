import { MapLike } from 'typescript'
import { replaceAll } from './string-utils'
import urljoin from 'url-join'
import { appendHash } from './dom-utils'

// This file shouldn't import anything as it is for exporting simple shared utility functions between various projects
export const EditorID = 'utopia-editor-root'
export const PortalTargetID = 'portal-target'
export const CanvasContextMenuPortalTargetID = 'canvas-contextmenu-portal-target'

export const RETURN_TO_PREPEND = 'return '

// eslint-disable-next-line @typescript-eslint/no-empty-function
export const NO_OP = () => {}

export type ValueOf<T> = T[keyof T]

export type PrimitiveType = number | string | boolean | bigint | symbol | null | undefined

export function identity<T>(t: T): T {
  return t
}

export function fastForEach<T>(a: readonly T[], fn: (t: T, index: number) => void) {
  for (var i = 0, len = a.length; i < len; i++) {
    if (i in a) {
      fn(a[i]!, i)
    }
  }
}

export function arrayEquals<T>(a: Array<T>, b: Array<T>, eq?: (l: T, r: T) => boolean): boolean {
  if (a === b) {
    return true
  } else {
    const equals = eq == null ? (l: T, r: T) => l === r : eq
    if (a.length === b.length) {
      for (let i = 0; i < a.length; i++) {
        if (!equals(a[i], b[i])) {
          return false
        }
      }
      return true
    } else {
      return false
    }
  }
}

export function objectEquals<T>(
  a: MapLike<T>,
  b: MapLike<T>,
  eq?: (l: T, r: T) => boolean,
): boolean {
  if (a === b) {
    return true
  } else {
    const equals = eq == null ? (l: T, r: T) => l === r : eq
    for (const aKey of Object.keys(a)) {
      if (aKey in b) {
        if (!equals(a[aKey], b[aKey])) {
          // Values in each object for the same key differ.
          return false
        }
      } else {
        // Value for key cannot be found in object 'b'.
        return false
      }
    }
    for (const bKey of Object.keys(b)) {
      if (!(bKey in a)) {
        // This key from 'b' isn't in 'a'.
        return false
      }
    }
    return true
  }
}

// Uses === as the semantics, which should be safe for the given types.
export function arrayContains<T extends string | boolean | number>(
  array: Array<T>,
  possibleEntry: T,
): boolean {
  return array.indexOf(possibleEntry) >= 0
}

export function longestCommonArray<T>(
  a: Array<T>,
  b: Array<T>,
  eq?: (l: T, r: T) => boolean,
): Array<T> {
  const equals = eq == null ? (l: T, r: T) => l === r : eq
  const commonMaxLength = Math.min(a.length, b.length)
  let commonArray: Array<T> = []
  for (let i = 0; i < commonMaxLength; i++) {
    if (equals(a[i], b[i])) {
      commonArray.push(a[i])
    } else {
      break
    }
  }
  return commonArray
}

export function getProjectLockedKey(projectId: string) {
  return `${projectId}-locked`
}

export function xor(a: boolean, b: boolean): boolean {
  return a != b
}

export const isBrowserEnvironment = process.env.JEST_WORKER_ID == undefined // If true this means we're not running in a Jest environment

export function urlSafeText(text: string): string {
  return encodeURIComponent(replaceAll(text.toLowerCase(), ' ', '-'))
}

function getProjectURLSuffix(projectId: string, projectName: string): string {
  return urlSafeText(`${projectId}-${projectName}`)
}

export function projectURLForProject(projectId: string, projectName: string): string {
  return `/p/${getProjectURLSuffix(projectId, projectName)}/`
}

export function shareURLForProject(prefix: string, projectId: string, projectName: string): string {
  return urljoin(prefix, `/share/${getProjectURLSuffix(projectId, projectName)}`)
}

export function unknownObjectProperty(o: unknown, key: string): any {
  if (typeof o === 'object' && o != null) {
    if (key in o) {
      // at this point we know `o` is an object and it has the prop, but Typescript threw an error
      return (o as any)[key]
    }
  }
  return undefined
}

export function createIframeUrl(base: string, assetName: string): string {
  return appendHash(urljoin(base, 'editor', assetName))
}
