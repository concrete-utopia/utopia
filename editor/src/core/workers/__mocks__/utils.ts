export function createTsWorker() {
  return null
}

export function createParserPrinterWorker() {
  return null
}

export function createLinterWorker() {
  return null
}

export function createWatchdogWorker() {
  return null
}
