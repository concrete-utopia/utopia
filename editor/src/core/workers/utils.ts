import { BASE_URL } from '../../common/env-vars'

/**
 * Webpack 5 only supports a single public path for assets and feeding the Worker url via `import.meta.url`
 * This causes problems when using a CDN, because of same origin restrictions, you are not allowed to load workers from cdns.
 *
 * The solution is to use the webpack-supported global __webpack_public_path__ variable, which updates the value returned by import.meta.url
 *
 * in all of these cases, oldPublicPath is the asset url (the CDN url in deployed editors). we temporarily change it to BASE_URL (which is the url the editor is loaded from)
 * and then after loading the worker, restore it to the CDN url.
 */

const WORKER_BASE_URL = `${BASE_URL}editor/`

import TSWorker from './ts/ts-worker?worker'
import ParserPrinterWorker from './parser-printer/parser-printer.worker?worker'
import LinterWorker from './linter/linter.worker?worker'
import WatchdogWorker from './watchdog.worker?worker'

export function createTsWorker(): Worker {
  // const oldPublicPath = __webpack_public_path__
  // __webpack_public_path__ = WORKER_BASE_URL
  // const worker = new Worker(new URL('./ts/ts.worker.ts', WORKER_BASE_URL))
  // __webpack_public_path__ = oldPublicPath
  // return worker
  return new TSWorker()
}

export function createParserPrinterWorker(): Worker {
  // const oldPublicPath = __webpack_public_path__
  // __webpack_public_path__ = WORKER_BASE_URL
  // const worker = new Worker(new URL('./parser-printer/parser-printer.worker.ts', WORKER_BASE_URL))
  // __webpack_public_path__ = oldPublicPath
  // return worker
  return new ParserPrinterWorker()
}

export function createLinterWorker(): Worker {
  // const oldPublicPath = __webpack_public_path__
  // __webpack_public_path__ = WORKER_BASE_URL
  // const worker = new Worker(new URL('./linter/linter.worker.ts', WORKER_BASE_URL))
  // __webpack_public_path__ = oldPublicPath
  // return worker
  return new LinterWorker()
}

export function createWatchdogWorker(): Worker {
  // const oldPublicPath = __webpack_public_path__
  // __webpack_public_path__ = WORKER_BASE_URL
  // const worker = new Worker(new URL('./watchdog.worker.ts', WORKER_BASE_URL))
  // __webpack_public_path__ = oldPublicPath
  // return worker
  return new WatchdogWorker()
}
