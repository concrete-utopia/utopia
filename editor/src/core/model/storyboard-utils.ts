import type { ProjectContentTreeRoot } from '../../components/assets'
import {
  addFileToProjectContents,
  getProjectFileByFilePath,
  walkContentsTree,
} from '../../components/assets'
import type { EditorModel } from '../../components/editor/action-types'
import type { EditorState } from '../../components/editor/store/editor-state'
import { StoryboardFilePath } from '../../components/editor/store/editor-state'
import type { Compare } from '../../utils/compare'
import {
  compareCompose,
  compareField,
  compareIfIs,
  compareOn,
  comparePrimitive,
} from '../../utils/compare'
import type { JSXElement } from '../shared/element-template'
import {
  emptyComments,
  isUtopiaJSXComponent,
  unparsedCode,
  UtopiaJSXComponent,
  utopiaJSXComponent,
} from '../shared/element-template'
import { forEachValue } from '../shared/object-utils'
import { forceNotNull } from '../shared/optional-utils'
import type { ParseSuccess } from '../shared/project-file-types'
import {
  EmptyExportsDetail,
  ExportDetail,
  exportVariable,
  exportVariables,
  forEachParseSuccess,
  importAlias,
  isParsedTextFile,
  isParseSuccess,
  isTextFile,
  mergeExportsDetail,
  parseSuccess,
  RevisionsState,
  textFile,
  textFileContents,
} from '../shared/project-file-types'
import { addImport } from '../workers/common/project-file-utils'
import {
  BakedInStoryboardUID,
  BakedInStoryboardVariableName,
  createSceneFromComponent,
  createStoryboardElement,
} from './scene-utils'

const PossiblyMainComponentNames: Array<string> = ['App', 'Application', 'Main']

interface DefaultComponentToImport {
  type: 'DEFAULT_COMPONENT_TO_IMPORT'
  path: string
}

interface NamedComponentToImport {
  type: 'NAMED_COMPONENT_TO_IMPORT'
  path: string
  possibleMainComponentName: boolean
  toImport: string
}

interface UnexportedRenderedComponent {
  type: 'UNEXPORTED_RENDERED_COMPONENT'
  path: string
  elementName: string
}

function defaultComponentToImport(path: string): DefaultComponentToImport {
  return {
    type: 'DEFAULT_COMPONENT_TO_IMPORT',
    path: path,
  }
}

function namedComponentToImport(
  path: string,
  possibleMainComponentName: boolean,
  toImport: string,
): NamedComponentToImport {
  return {
    type: 'NAMED_COMPONENT_TO_IMPORT',
    path: path,
    possibleMainComponentName: possibleMainComponentName,
    toImport: toImport,
  }
}

function unexportedRenderedComponent(
  path: string,
  elementName: string,
): UnexportedRenderedComponent {
  return {
    type: 'UNEXPORTED_RENDERED_COMPONENT',
    path: path,
    elementName: elementName,
  }
}

type ComponentToImport =
  | DefaultComponentToImport
  | NamedComponentToImport
  | UnexportedRenderedComponent

function isDefaultComponentToImport(
  toImport: ComponentToImport,
): toImport is DefaultComponentToImport {
  return toImport.type === 'DEFAULT_COMPONENT_TO_IMPORT'
}

function isNamedComponentToImport(toImport: ComponentToImport): toImport is NamedComponentToImport {
  return toImport.type === 'NAMED_COMPONENT_TO_IMPORT'
}

function isUnexportedRenderedComponent(
  toImport: ComponentToImport,
): toImport is UnexportedRenderedComponent {
  return toImport.type === 'UNEXPORTED_RENDERED_COMPONENT'
}

const compareDefaultComponentToImport: Compare<DefaultComponentToImport> = compareCompose(
  compareOn((toImport) => toImport.path.split('/').length, comparePrimitive),
)

const compareNamedComponentToImport: Compare<NamedComponentToImport> = compareCompose(
  compareOn((toImport) => toImport.path.split('/').length, comparePrimitive),
  compareField('possibleMainComponentName', comparePrimitive),
  compareField('toImport', comparePrimitive),
)

const compareUnexportedRenderedComponent: Compare<UnexportedRenderedComponent> = compareCompose(
  compareOn((toImport) => toImport.path.split('/').length, comparePrimitive),
  compareOn(
    (toImport) => PossiblyMainComponentNames.includes(toImport.elementName),
    comparePrimitive,
  ),
)

// Highest priority last.
const componentToImportTypesInPriorityOrder: Array<ComponentToImport['type']> = [
  'NAMED_COMPONENT_TO_IMPORT',
  'DEFAULT_COMPONENT_TO_IMPORT',
  'UNEXPORTED_RENDERED_COMPONENT',
]

const compareComponentToImport: Compare<ComponentToImport> = compareCompose(
  compareOn((toImport) => toImport.path.split('/').length, comparePrimitive),
  compareIfIs(isDefaultComponentToImport, compareDefaultComponentToImport),
  compareIfIs(isNamedComponentToImport, compareNamedComponentToImport),
  compareIfIs(isUnexportedRenderedComponent, compareUnexportedRenderedComponent),
  compareOn(
    (toImport) => componentToImportTypesInPriorityOrder.indexOf(toImport.type),
    comparePrimitive,
  ),
)

export function addStoryboardFileToProject(editorModel: EditorModel): EditorModel | null {
  const storyboardFile = getProjectFileByFilePath(editorModel.projectContents, StoryboardFilePath)
  if (storyboardFile == null) {
    let currentImportCandidate: ComponentToImport | null = null
    function updateCandidate(importCandidate: ComponentToImport): void {
      if (currentImportCandidate == null) {
        currentImportCandidate = importCandidate
      } else {
        if (compareComponentToImport(importCandidate, currentImportCandidate) > 0) {
          currentImportCandidate = importCandidate
        }
      }
    }
    walkContentsTree(editorModel.projectContents, (fullPath, file) => {
      if (isParsedTextFile(file)) {
        // For those successfully parsed files, we want to search all of the components.
        forEachParseSuccess((success) => {
          let namedExportKeys: Array<string> = []
          for (const exportDetail of success.exportsDetail) {
            switch (exportDetail.type) {
              case 'EXPORT_DEFAULT_FUNCTION_OR_CLASS':
                updateCandidate(defaultComponentToImport(fullPath))
                break
              case 'EXPORT_CLASS':
                {
                  const possibleMainComponentName = PossiblyMainComponentNames.includes(
                    exportDetail.className,
                  )
                  updateCandidate(
                    namedComponentToImport(
                      fullPath,
                      possibleMainComponentName,
                      exportDetail.className,
                    ),
                  )
                }
                break
              case 'EXPORT_FUNCTION':
                {
                  const possibleMainComponentName = PossiblyMainComponentNames.includes(
                    exportDetail.functionName,
                  )
                  updateCandidate(
                    namedComponentToImport(
                      fullPath,
                      possibleMainComponentName,
                      exportDetail.functionName,
                    ),
                  )
                }
                break
              case 'EXPORT_VARIABLES':
              case 'EXPORT_DESTRUCTURED_ASSIGNMENT':
              case 'REEXPORT_VARIABLES':
                {
                  for (const exportVar of exportDetail.variables) {
                    const exportName = exportVar.variableAlias ?? exportVar.variableName
                    if (exportName !== 'default') {
                      namedExportKeys.push(exportName)
                      const possibleMainComponentName =
                        PossiblyMainComponentNames.includes(exportName)
                      updateCandidate(
                        namedComponentToImport(fullPath, possibleMainComponentName, exportName),
                      )
                    }
                  }
                }
                break
              case 'EXPORT_IDENTIFIER':
                {
                  const possibleMainComponentName = PossiblyMainComponentNames.includes(
                    exportDetail.name,
                  )
                  updateCandidate(
                    namedComponentToImport(fullPath, possibleMainComponentName, exportDetail.name),
                  )
                }
                break
              case 'REEXPORT_WILDCARD':
                break
              case 'EXPORT_VARIABLES_WITH_MODIFIER':
                {
                  for (const exportName of exportDetail.variables) {
                    if (exportName !== 'default') {
                      namedExportKeys.push(exportName)
                      const possibleMainComponentName =
                        PossiblyMainComponentNames.includes(exportName)
                      updateCandidate(
                        namedComponentToImport(fullPath, possibleMainComponentName, exportName),
                      )
                    }
                  }
                }
                break
              default:
                const _exhaustiveCheck: never = exportDetail
                throw new Error(`Unhandled type ${JSON.stringify(exportDetail)}`)
            }
          }
        }, file.fileContents.parsed)
      }
    })

    if (currentImportCandidate == null) {
      return null
    } else {
      return addStoryboardFileForComponent(currentImportCandidate, editorModel)
    }
  } else {
    return null
  }
}

function addStoryboardFileForComponent(
  createFileWithComponent: ComponentToImport,
  editorModel: EditorState,
): EditorState {
  // Add import of storyboard and scene components.
  let imports = addImport(StoryboardFilePath, 'react', null, [], 'React', {})
  imports = addImport(
    StoryboardFilePath,
    'utopia-api',
    null,
    [importAlias('Storyboard'), importAlias('Scene')],
    null,
    imports,
  )
  // Create the storyboard variable.
  let sceneElement: JSXElement
  let updatedProjectContents: ProjectContentTreeRoot = editorModel.projectContents
  switch (createFileWithComponent.type) {
    case 'NAMED_COMPONENT_TO_IMPORT':
      sceneElement = createSceneFromComponent(
        StoryboardFilePath,
        createFileWithComponent.toImport,
        'scene-1',
      )
      imports = addImport(
        StoryboardFilePath,
        createFileWithComponent.path,
        null,
        [importAlias(createFileWithComponent.toImport)],
        null,
        imports,
      )
      break
    case 'DEFAULT_COMPONENT_TO_IMPORT':
      sceneElement = createSceneFromComponent(StoryboardFilePath, 'StoryboardComponent', 'scene-1')
      imports = addImport(
        StoryboardFilePath,
        createFileWithComponent.path,
        'StoryboardComponent',
        [],
        null,
        imports,
      )
      break
    case 'UNEXPORTED_RENDERED_COMPONENT':
      sceneElement = createSceneFromComponent(
        StoryboardFilePath,
        createFileWithComponent.elementName,
        'scene-1',
      )
      imports = addImport(
        StoryboardFilePath,
        createFileWithComponent.path,
        null,
        [importAlias(createFileWithComponent.elementName)],
        null,
        imports,
      )
      // Modify the targeted file to export the component we're interested in.
      const fileToModify = forceNotNull(
        `Unable to find file at ${createFileWithComponent.path}`,
        getProjectFileByFilePath(updatedProjectContents, createFileWithComponent.path),
      )
      if (isTextFile(fileToModify)) {
        if (isParseSuccess(fileToModify.fileContents.parsed)) {
          const currentSuccess: ParseSuccess = fileToModify.fileContents.parsed
          const updatedExports = mergeExportsDetail(currentSuccess.exportsDetail, [
            exportVariables([exportVariable(createFileWithComponent.elementName, null)]),
          ])
          const updatedParseSuccess = parseSuccess(
            currentSuccess.imports,
            currentSuccess.topLevelElements,
            currentSuccess.highlightBounds,
            currentSuccess.jsxFactoryFunction,
            currentSuccess.combinedTopLevelArbitraryBlock,
            updatedExports,
            currentSuccess.fullHighlightBounds,
          )
          const updatedContents = textFileContents(
            fileToModify.fileContents.code,
            updatedParseSuccess,
            RevisionsState.ParsedAhead,
          )
          const updatedTextFile = textFile(
            updatedContents,
            fileToModify.lastSavedContents,
            updatedParseSuccess,
            fileToModify.versionNumber + 1,
          )
          updatedProjectContents = addFileToProjectContents(
            updatedProjectContents,
            createFileWithComponent.path,
            updatedTextFile,
          )
        } else {
          throw new Error(`Unexpectedly ${createFileWithComponent.path} is not a parse success.`)
        }
      } else {
        throw new Error(`${createFileWithComponent.path} was not a text file as expected.`)
      }
      break
    default:
      const _exhaustiveCheck: never = createFileWithComponent
      throw new Error(`Unhandled type ${JSON.stringify(createFileWithComponent)}`)
  }
  const storyboardElement = createStoryboardElement([sceneElement], BakedInStoryboardUID)
  const storyboardComponent = utopiaJSXComponent(
    BakedInStoryboardVariableName,
    false,
    'var',
    'block',
    null,
    [],
    storyboardElement,
    null,
    false,
    emptyComments,
  )
  // Add the component import.
  // Create the file.
  const success = parseSuccess(
    imports,
    [unparsedCode('\n\n'), storyboardComponent],
    {},
    null,
    null,
    [exportVariables([exportVariable(BakedInStoryboardVariableName, null)])],
    {},
  )
  const storyboardFileContents = textFile(
    textFileContents('', success, RevisionsState.ParsedAhead),
    null,
    success,
    0,
  )

  // Update the model.
  updatedProjectContents = addFileToProjectContents(
    updatedProjectContents,
    StoryboardFilePath,
    storyboardFileContents,
  )
  return {
    ...editorModel,
    projectContents: updatedProjectContents,
  }
}
