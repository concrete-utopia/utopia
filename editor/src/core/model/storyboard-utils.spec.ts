import {
  addFileToProjectContents,
  getContentsTreeFileFromString,
  removeFromProjectContents,
} from '../../components/assets'
import {
  editorModelFromPersistentModel,
  EditorState,
  StoryboardFilePath,
} from '../../components/editor/store/editor-state'
import { complexDefaultProject } from '../../sample-projects/sample-project-utils'
import { clearTopLevelElementUniqueIDs } from '../shared/element-template'
import {
  foldParsedTextFile,
  isParseSuccess,
  isTextFile,
  ParsedTextFile,
  RevisionsState,
  textFile,
  textFileContents,
  unparsed,
} from '../shared/project-file-types'
import { emptySet } from '../shared/set-utils'
import { NO_OP } from '../shared/utils'
import { lintAndParse } from '../workers/parser-printer/parser-printer'
import { addStoryboardFileToProject } from './storyboard-utils'

function createTestProjectLackingStoryboardFile(): EditorState {
  const appFile = `
import React from 'react'
export var App = (props) => {
  return <div style={{ ...props.style}} data-uid={'aaa'} data-label={'Hat'} />
}`
  const baseModel = complexDefaultProject()
  const parsedFile = lintAndParse(StoryboardFilePath, appFile, null, emptySet()) as ParsedTextFile

  if (!isParseSuccess(parsedFile)) {
    throw new Error('The test file parse failed')
  }

  const projectContentsWithoutStoryboard = removeFromProjectContents(
    baseModel.projectContents,
    StoryboardFilePath,
  )

  const persistentModel = {
    ...baseModel,
    projectContents: addFileToProjectContents(
      projectContentsWithoutStoryboard,
      '/src/app.js',
      textFile(
        textFileContents(appFile, parsedFile, RevisionsState.BothMatch),
        null,
        parsedFile,
        Date.now(),
      ),
    ),
  }

  return editorModelFromPersistentModel(persistentModel, NO_OP)
}

describe('addStoryboardFileToProject', () => {
  it('adds storyboard file to project that does not have one', () => {
    const actualResult = addStoryboardFileToProject(createTestProjectLackingStoryboardFile())
    if (actualResult == null) {
      throw new Error('Editor state was not updated.')
    } else {
      const storyboardFile = getContentsTreeFileFromString(
        actualResult.projectContents,
        StoryboardFilePath,
      )
      if (storyboardFile == null) {
        throw new Error('No storyboard file was created.')
      } else {
        if (isTextFile(storyboardFile)) {
          const topLevelElements = foldParsedTextFile(
            (_) => [],
            (success) => {
              return success.topLevelElements.map(clearTopLevelElementUniqueIDs)
            },
            (_) => [],
            storyboardFile.fileContents.parsed,
          )
          expect(topLevelElements).toMatchSnapshot()
        } else {
          throw new Error('Storyboard file is not a UI JS file.')
        }
      }
    }
  })
  it('does not add a storyboard file to a project that already has one', () => {
    const expectedFile = textFile(
      textFileContents(
        'oh no, this is not a real storyboard file',
        unparsed,
        RevisionsState.CodeAhead,
      ),
      null,
      null,
      0,
    )
    let editorModel = createTestProjectLackingStoryboardFile()
    editorModel = {
      ...editorModel,
      projectContents: addFileToProjectContents(
        editorModel.projectContents,
        StoryboardFilePath,
        expectedFile,
      ),
    }
    const actualResult = addStoryboardFileToProject(editorModel)
    expect(actualResult).toBeNull()
  })
})
