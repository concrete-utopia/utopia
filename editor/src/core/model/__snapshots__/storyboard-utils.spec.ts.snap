// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`addStoryboardFileToProject adds storyboard file to project that does not have one 1`] = `
Array [
  Object {
    "rawCode": "

",
    "type": "UNPARSED_CODE",
  },
  Object {
    "arbitraryJSBlock": null,
    "blockOrExpression": "block",
    "declarationSyntax": "const",
    "functionWrapping": Array [],
    "isFunction": false,
    "name": "storyboard",
    "params": null,
    "propsUsed": Array [],
    "returnStatementComments": Object {
      "leadingComments": Array [],
      "trailingComments": Array [],
    },
    "rootElement": Object {
      "children": Array [
        Object {
          "children": Array [
            Object {
              "children": Array [],
              "name": Object {
                "baseVariable": "App",
                "propertyPath": Object {
                  "propertyElements": Array [],
                },
              },
              "props": Array [
                Object {
                  "comments": Object {
                    "leadingComments": Array [],
                    "trailingComments": Array [],
                  },
                  "key": "data-uid",
                  "type": "JSX_ATTRIBUTES_ENTRY",
                  "value": Object {
                    "comments": Object {
                      "leadingComments": Array [],
                      "trailingComments": Array [],
                    },
                    "type": "ATTRIBUTE_VALUE",
                    "uid": "",
                    "value": "9d2",
                  },
                },
              ],
              "type": "JSX_ELEMENT",
              "uid": "",
            },
          ],
          "name": Object {
            "baseVariable": "Scene",
            "propertyPath": Object {
              "propertyElements": Array [],
            },
          },
          "props": Array [
            Object {
              "comments": Object {
                "leadingComments": Array [],
                "trailingComments": Array [],
              },
              "key": "data-uid",
              "type": "JSX_ATTRIBUTES_ENTRY",
              "value": Object {
                "comments": Object {
                  "leadingComments": Array [],
                  "trailingComments": Array [],
                },
                "type": "ATTRIBUTE_VALUE",
                "uid": "",
                "value": "scene-1",
              },
            },
            Object {
              "comments": Object {
                "leadingComments": Array [],
                "trailingComments": Array [],
              },
              "key": "style",
              "type": "JSX_ATTRIBUTES_ENTRY",
              "value": Object {
                "comments": Object {
                  "leadingComments": Array [],
                  "trailingComments": Array [],
                },
                "type": "ATTRIBUTE_VALUE",
                "uid": "",
                "value": Object {
                  "height": 812,
                  "left": 0,
                  "position": "absolute",
                  "top": 0,
                  "width": 375,
                },
              },
            },
          ],
          "type": "JSX_ELEMENT",
          "uid": "",
        },
      ],
      "name": Object {
        "baseVariable": "Storyboard",
        "propertyPath": Object {
          "propertyElements": Array [],
        },
      },
      "props": Array [
        Object {
          "comments": Object {
            "leadingComments": Array [],
            "trailingComments": Array [],
          },
          "key": "data-uid",
          "type": "JSX_ATTRIBUTES_ENTRY",
          "value": Object {
            "comments": Object {
              "leadingComments": Array [],
              "trailingComments": Array [],
            },
            "type": "ATTRIBUTE_VALUE",
            "uid": "",
            "value": "utopia-storyboard-uid",
          },
        },
      ],
      "type": "JSX_ELEMENT",
      "uid": "",
    },
    "type": "UTOPIA_JSX_COMPONENT",
    "usedInReactDOMRender": false,
  },
]
`;
